datastax 4.4.0 driver

pollend opened the pull request:
This is a work in progress with added datastax driver for 4.4.0. They have a entity binding library similar to JPA so I think that would also be relevant if that was to be moving to a separate sub-project. 

I'm not sure what the plan is regards to this library but I'm starting to put together a configuration for the driver. Looks like they have moved over to using `typesafe.config` so I'm trying to map the properties from micronaut into the config. I'm not sure if this is the best way to approach the problem but it seems to be somewhat working for the most part. It's just a matter of implementing everything as specified. I'll have to also add some interfaces to generalize and hopefully line this up more with code standards. 

I've also been looking at this project for managing migrations, but this is a separate concern: https://github.com/hhandoko/cassandra-migration

I found the recommend properties from this reference: https://github.com/datastax/java-driver/blob/4.0.1/core/src/main/resources/reference.conf

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/micronaut-projects/micronaut-cassandra?pullRequest=7) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/micronaut-projects/micronaut-cassandra?pullRequest=7) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/micronaut-projects/micronaut-cassandra?pullRequest=7) it.</sub>

willbuck commented:
I will take a look at this tomorrow @pollend, I am not super familiar with Cassandra or Datastax but I had been looking at this a week or so ago and very much appreciate your interest in tackling the upgrade! It certainly wasn't trivial enough for dependabot to knock out :)

pollend commented:
@willbuck awesome, I made some more tweaks and added the two missing test cases from the original. The coverage should be better and what do you think about the dao mapping they have. Would this live under this repo If I were to add that at some point?

https://docs.datastax.com/en/developer/java-driver/4.4/manual/mapper/daos/

willbuck commented on cassandra/src/test/groovy/io/micronaut/configuration/cassandra/CassandraConfigurationSpec.groovy:
```diff
@@ -35,86 +39,78 @@ class CassandraConfigurationSpec extends Specification {
-
-        cleanup:
+        then:
+        EmbeddedCassandraServerHelper.cleanEmbeddedCassandra()
```
This is a cool utility, I'm glad you were aware of it!

willbuck commented on cassandra/src/test/groovy/io/micronaut/configuration/cassandra/CassandraConfigurationSpec.groovy:
```diff
@@ -35,86 +39,78 @@ class CassandraConfigurationSpec extends Specification {
 
     void "test single cluster connection"() {
         given:
-        // tag::single[]
```
These tags link up to the documentation; are you planning to replace them / edit the docs to better illustrate new usage?

pollend replied:
oh, didn't realize that. I guess I'll have to add that back in

willbuck replied:
Not a problem! Thanks :)

willbuck commented on cassandra/src/test/groovy/io/micronaut/configuration/cassandra/CassandraConfigurationSpec.groovy:
```diff
@@ -35,86 +39,78 @@ class CassandraConfigurationSpec extends Specification {
-                 'cassandra.default.port': 9042,
-                 'cassandra.default.maxSchemaAgreementWaitSeconds': 20,
-                 'cassandra.default.ssl': true]
+                ['cassandra.datasource.default.basic.contact-points'                        : ["localhost:9142"],
```
So with this upgrade, old configs would break entirely? Is that accurate? One thing I was thinking about was adding some sort of mapping functions to graft the old config onto the new values: do you think that's doable?

willbuck commented on cassandra/src/test/groovy/io/micronaut/configuration/cassandra/CassandraConfigurationSpec.groovy:
```diff
@@ -35,86 +39,78 @@ class CassandraConfigurationSpec extends Specification {
 
     void "test multiple cluster connections"() {
         given:
-        // tag::multiple[]
```
Same here as the `// tag::single[]`, this is linked in the documentation and we'll want to be aware of providing an alternative or keeping it in and modifying the surrounding text to support the new config.

willbuck commented on cassandra/src/main/java/io/micronaut/configuration/cassandra/CassandraConfiguration.java:
```diff
@@ -26,21 +25,25 @@
-@EachProperty(value = "cassandra", primary = "default")
+@EachProperty(value = CassandraConfiguration.PREFIX, primary = "default")
 public class CassandraConfiguration {
+    public static final String PREFIX = "cassandra.datasource";
```
Is this a value you've chosen, or one dictated by the new driver?

pollend replied:
It's a value that I choose but we can use something better if that is more appropriate. I was looking at the datastax document and I was thinking of maybe adding cassandra.dao, cassandra.mapping etc ...

willbuck commented on cassandra/src/main/java/io/micronaut/configuration/cassandra/CassandraConfiguration.java:
```diff
@@ -26,21 +25,25 @@
+    private final String name;
 
-    @SuppressWarnings("WeakerAccess")
-    @ConfigurationBuilder(allowZeroArgs = true, prefixes = { "with", "add" })
```
Perhaps I'm missing something; what's replacing this annotation as a means of building the connection details from the micronaut config (like `application.yml`)

pollend replied:
so they don't have an object that you can just map the configs to easily. Before this I had a large amount of static classes that I was trying to map configuration to but that posed too much of a challenge.  I manually resolve that in the CassandraSessionFactory where I use the cassandra.datasource.<name>.<config_path> where config_path is a path define in DefaultDriverOption. The values I chose for the default are from this reference: https://github.com/datastax/java-driver/blob/4.0.1/core/src/main/resources/reference.conf

the datastax driver 4.4.0 is using this to define configurations: https://github.com/lightbend/config

willbuck replied:
Are there comparable properties to the ones that would have been provided previously (like `cassandra.default.port` and the like provided in the tests?)

pollend replied:
actually, no I don't think so. Any suggestions how I would go about this, I can move it back to the old path and add some properties so the cassandra driver would match. I'm not even sure if this will drop into place with the old driver. It looks like the whole thing was rewritten from 3 to 4. I can check those old paths and remap those properties, but wouldn't cassandra.datasource map back to the old config with cassandra.default.

willbuck commented:
Thank you for your efforts on this @pollend ! For a brief primer on the docs:

1. They are asciidoc format (pretty simple, like markdown but with some differences) [The reference sheet on their site is pretty comprehensive](https://asciidoctor.org/docs/asciidoc-syntax-quick-reference/), minus some custom extensions we supply as a part of the [micronaut-docs](https://github.com/micronaut-projects/micronaut-docs) project.
2. You can generate them using `./gradlew docs`, and then view them under `build/docs/index.html`

Feel free to reach out with any further questions!

pollend commented:
I pretty much have all the properties that could be used to configure the driver so I'm thinking I only need to list out the more relevant ones right?

pollend commented:
I was also curious how this 4.4.0 version of this driver stack up against 3.3.0. There are a lot more knobs to tweak and adjust then there was in version 3? I'm still new to micronaut but the way I'm querying for properties, is it compatible with the yml configuration?

pollend commented:
@willbuck I decided to move the prefix back to the old path. I guess I could just map the old paths back to the new one manually but the driver changes a lot. You pretty much need to rework the config and test it because the default configuration might not be ideal. What other test should be run to test this version of driver?

willbuck commented:
Hey @pollend! I was speaking some with @jameskleeh and @graemerocher and we don't think we'll have the resources to maintain a compatability layer like the one I'm describing, so they suggested instead that we bump up the major version of this library with the driver upgrade, and simply make note in the docs that there is a breaking change in how the config is structured (due to the driver changing). 

So I would suggest the following:
1. Create a `changelog.adoc` file and add it to the `toc.yml` in between `introduction` and `setup`
2. Make note of the changes under a `2.0.0` heading of how folks will need to change their configs to utilize the new driver (if Datastax / Cassandra have a guide somewhere already, a link to it would be sufficient, perhaps with a little snippet of the new tests pulled in)
3. Make note of folks wishing to keep their existing config to remain on the `1.2.x` line of the library.

Does that make sense?

pollend commented:
yep, I agree with those suggestions. I'll work on that in the next couple of days and update you when I finish.

pollend commented:
so some of the documentation has information about the fields. This is the list of properties that will be supported with this version of the drive. The information about the field is [here](https://github.com/datastax/java-driver/blob/4.x/core/src/main/java/com/datastax/oss/driver/api/core/config/DefaultDriverOption.java). The field with the javadocs provides information about the what the fiel does and the datatype that is accepted.  Any clue about generating a table with all of these?
```
| basic.contact-points | String | Default 
| cassandra.*.basic.session-name | String | Default 
| cassandra.*.basic.session-keyspace | String | Default 
| cassandra.*.basic.config-reload-interval | String | Default 
| cassandra.*.basic.request.timeout | String | Default 
| cassandra.*.basic.request.consistency | String | Default 
| cassandra.*.basic.request.page-size | String | Default 
| cassandra.*.basic.request.serial-consistency | String | Default 
| cassandra.*.basic.request.default-idempotence | String | Default 
| cassandra.*.basic.load-balancing-policy | String | Default 
| cassandra.*.basic.load-balancing-policy.class | String | Default 
| cassandra.*.basic.load-balancing-policy.local-datacenter | String | Default 
| cassandra.*.basic.load-balancing-policy.filter.class | String | Default 
| cassandra.*.advanced.connection.init-query-timeout | String | Default 
| cassandra.*.advanced.connection.set-keyspace-timeout | String | Default 
| cassandra.*.advanced.connection.max-requests-per-connection | String | Default 
| cassandra.*.advanced.connection.max-orphan-requests | String | Default 
| cassandra.*.advanced.connection.warn-on-init-error | String | Default 
| cassandra.*.advanced.connection.pool.local.size | String | Default 
| cassandra.*.advanced.connection.pool.remote.size | String | Default 
| cassandra.*.advanced.reconnect-on-init | String | Default 
| cassandra.*.advanced.reconnection-policy.class | String | Default 
| cassandra.*.advanced.reconnection-policy.base-delay | String | Default 
| cassandra.*.advanced.reconnection-policy.max-delay | String | Default 
| cassandra.*.advanced.retry-policy | String | Default 
| cassandra.*.advanced.retry-policy.class | String | Default 
| cassandra.*.advanced.speculative-execution-policy | String | Default 
| cassandra.*.advanced.speculative-execution-policy.class | String | Default 
| cassandra.*.advanced.speculative-execution-policy.max-executions | String | Default 
| cassandra.*.advanced.speculative-execution-policy.delay | String | Default 
| cassandra.*.advanced.auth-provider.class | String | Default 
| cassandra.*.advanced.auth-provider.username | String | Default 
| cassandra.*.advanced.auth-provider.password | String | Default 
| cassandra.*.advanced.ssl-engine-factory.class | String | Default 
| cassandra.*.advanced.ssl-engine-factory.cipher-suites | String | Default 
| cassandra.*.advanced.ssl-engine-factory.hostname-validation | String | Default 
| cassandra.*.advanced.ssl-engine-factory.keystore-path | String | Default 
| cassandra.*.advanced.ssl-engine-factory.keystore-password | String | Default 
| cassandra.*.advanced.ssl-engine-factory.truststore-path | String | Default 
| cassandra.*.advanced.ssl-engine-factory.truststore-password | String | Default 
| cassandra.*.advanced.timestamp-generator.class | String | Default 
| cassandra.*.advanced.timestamp-generator.force-java-clock | String | Default 
| cassandra.*.advanced.timestamp-generator.drift-warning.interval | String | Default 
| cassandra.*.advanced.request-tracker.class | String | Default 
| cassandra.*.advanced.request-tracker.logs.success.enabled | String | Default 
| cassandra.*.advanced.request-tracker.logs.slow.threshold | String | Default 
| cassandra.*.advanced.request-tracker.logs.slow.enabled | String | Default 
| cassandra.*.advanced.request-tracker.logs.error.enabled | String | Default 
| cassandra.*.advanced.request-tracker.logs.max-query-length | String | Default 
| cassandra.*.advanced.request-tracker.logs.show-values | String | Default 
| cassandra.*.advanced.request-tracker.logs.max-value-length | String | Default 
| cassandra.*.advanced.request-tracker.logs.max-values | String | Default 
| cassandra.*.advanced.request-tracker.logs.show-stack-traces | String | Default 
| cassandra.*.advanced.throttler.class | String | Default 
| cassandra.*.advanced.throttler.max-concurrent-requests | String | Default 
| cassandra.*.advanced.throttler.max-requests-per-second | String | Default 
| cassandra.*.advanced.throttler.max-queue-size | String | Default 
| cassandra.*.advanced.throttler.drain-interval | String | Default 
| cassandra.*.advanced.node-state-listener.class | String | Default 
| cassandra.*.advanced.schema-change-listener.class | String | Default 
| cassandra.*.advanced.address-translator.class | String | Default 
| cassandra.*.advanced.protocol.version | String | Default 
| cassandra.*.advanced.protocol.compression | String | Default 
| cassandra.*.advanced.protocol.max-frame-length | String | Default 
| cassandra.*.advanced.request.warn-if-set-keyspace | String | Default 
| cassandra.*.advanced.request.trace.attempts | String | Default 
| cassandra.*.advanced.request.trace.interval | String | Default 
| cassandra.*.advanced.request.trace.consistency | String | Default 
| cassandra.*.advanced.metrics.session.enabled | String | Default 
| cassandra.*.advanced.metrics.node.enabled | String | Default 
| cassandra.*.advanced.metrics.session.cql-requests.highest-latency | String | Default 
| cassandra.*.advanced.metrics.session.cql-requests.significant-digits | String | Default 
| cassandra.*.advanced.metrics.session.cql-requests.refresh-interval | String | Default 
| cassandra.*.advanced.metrics.session.throttling.delay.highest-latency | String | Default 
| cassandra.*.advanced.metrics.session.throttling.delay.significant-digits | String | Default 
| cassandra.*.advanced.metrics.session.throttling.delay.refresh-interval | String | Default 
| cassandra.*.advanced.metrics.node.cql-messages.highest-latency | String | Default 
| cassandra.*.advanced.metrics.node.cql-messages.significant-digits | String | Default 
| cassandra.*.advanced.metrics.node.cql-messages.refresh-interval | String | Default 
| cassandra.*.advanced.socket.tcp-no-delay | String | Default 
| cassandra.*.advanced.socket.keep-alive | String | Default 
| cassandra.*.advanced.socket.reuse-address | String | Default 
| cassandra.*.advanced.socket.linger-interval | String | Default 
| cassandra.*.advanced.socket.receive-buffer-size | String | Default 
| cassandra.*.advanced.socket.send-buffer-size | String | Default 
| cassandra.*.advanced.heartbeat.interval | String | Default 
| cassandra.*.advanced.heartbeat.timeout | String | Default 
| cassandra.*.advanced.metadata.topology-event-debouncer.window | String | Default 
| cassandra.*.advanced.metadata.topology-event-debouncer.max-events | String | Default 
| cassandra.*.advanced.metadata.schema.enabled | String | Default 
| cassandra.*.advanced.metadata.schema.request-timeout | String | Default 
| cassandra.*.advanced.metadata.schema.request-page-size | String | Default 
| cassandra.*.advanced.metadata.schema.refreshed-keyspaces | String | Default 
| cassandra.*.advanced.metadata.schema.debouncer.window | String | Default 
| cassandra.*.advanced.metadata.schema.debouncer.max-events | String | Default 
| cassandra.*.advanced.metadata.token-map.enabled | String | Default 
| cassandra.*.advanced.control-connection.timeout | String | Default 
| cassandra.*.advanced.control-connection.schema-agreement.interval | String | Default 
| cassandra.*.advanced.control-connection.schema-agreement.timeout | String | Default 
| cassandra.*.advanced.control-connection.schema-agreement.warn-on-failure | String | Default 
| cassandra.*.advanced.prepared-statements.prepare-on-all-nodes | String | Default 
| cassandra.*.advanced.prepared-statements.reprepare-on-up.enabled | String | Default 
| cassandra.*.advanced.prepared-statements.reprepare-on-up.check-system-table | String | Default 
| cassandra.*.advanced.prepared-statements.reprepare-on-up.max-statements | String | Default 
| cassandra.*.advanced.prepared-statements.reprepare-on-up.max-parallelism | String | Default 
| cassandra.*.advanced.prepared-statements.reprepare-on-up.timeout | String | Default 
| cassandra.*.advanced.netty.io-group.size | String | Default 
| cassandra.*.advanced.netty.io-group.shutdown.quiet-period | String | Default 
| cassandra.*.advanced.netty.io-group.shutdown.timeout | String | Default 
| cassandra.*.advanced.netty.io-group.shutdown.unit | String | Default 
| cassandra.*.advanced.netty.admin-group.size | String | Default 
| cassandra.*.advanced.netty.admin-group.shutdown.quiet-period | String | Default 
| cassandra.*.advanced.netty.admin-group.shutdown.timeout | String | Default 
| cassandra.*.advanced.netty.admin-group.shutdown.unit | String | Default 
| cassandra.*.advanced.coalescer.max-runs-with-no-work | String | Default 
| cassandra.*.advanced.coalescer.reschedule-interval | String | Default 
| cassandra.*.advanced.resolve-contact-points | String | Default 
| cassandra.*.advanced.netty.timer.tick-duration | String | Default 
| cassandra.*.advanced.netty.timer.ticks-per-wheel | String | Default 
| cassandra.*.advanced.request.log-warnings | String | Default 
| cassandra.*.advanced.netty.daemon | String | Default 
| cassandra.*.basic.cloud.secure-connect-bundle | String | Default 
| cassandra.*.basic.load-balancing-policy.slow-replica-avoidance | | | |
```

pollend commented:
So I added a section with additional information how the datastax driver configurations map to micronaut. I moved the changelog into introduction which matches up with a couple of the other documented libraries.

willbuck commented on src/main/docs/guide/additional.adoc:
```diff
@@ -0,0 +1,25 @@
+datastax is configured using https://github.com/lightbend/config[lightbend/config] under the field `datastax-java-driver.\*`. the equivalent bean is mapped under `cassandra.*.*`. fields provided under
+application.conf will act as fallback properties when building https://docs.datastax.com/en/drivers/java/4.4/com/datastax/oss/driver/api/core/CqlSession.html[CqlSession].
+
+.datastax configuration under `lightbend/confg` - application.conf
```
I think this should probably be `lightbend/config`, yea?
```suggestion
.datastax configuration under `lightbend/config` - application.conf
```

willbuck commented on src/main/docs/guide/introduction.adoc:
```diff
@@ -1 +1,16 @@
 The `micronaut-cassandra` module includes support for integrating Micronaut services with Cassandra.
+
+=== Release History
```
This is a good fit and you're right, didn't necessarily need to be in its own `changelog.adoc` file

willbuck commented:
> Any clue how I would go about generating this?

Just to make sure I understand what you're asking, you're suggesting an automated script that would:
- Parse the `DefaultDriverConfig.java` file
- Generate a table of all the keys
- Add to each row the type of each key (`String`, `int`, `URL` etc)
- Add to each row what the default value is
- Parse the javadocs comments above each key and append that to each row as a "Description" 

Something like that?

If so, I can't say I have a genius idea to make such a script easy, but perhaps a Groovy AST visitor of some kind would be capable of that?

willbuck commented:
@pollend my colleague @jameskleeh was kind enough to point out that the thing you're looking for does, in fact, exist already! It simply wasn't being used in the project. Pull in master to your branch, as my latest commit adds `annotationProcessor "io.micronaut.docs:micronaut-docs-asciidoc-config-props:$micronautDocsVersion"` to the `build.gradle`. If you run `./gradlew docs` and check out the `index.html` file created, you will see `Conf` along the top nav bar, and it should generate what you're looking for!

pollend commented:
here is the sample project I put together: https://github.com/pollend/micronaut-cassandra-sample

so CQLsession will thrown an exception if the cassandra instance isn't quite ready. not sure If this is the correct way to approach this with a while loop and a thread sleep. Is there a mechanism already built into micronaut that can be added to handle when the connection fails?

pollend commented:
here is a sample project with the driver: https://github.com/pollend/micronaut-cassandra-sample

pollend commented:
@willbuck I think those docs are generated from the annotations. I've manually built the configurations from the driver and I've basically just mapped field to field. you can see that in CqlSessionBuilder.

jameskleeh commented on cassandra/src/main/java/io/micronaut/configuration/cassandra/CassandraSessionFactory.java:
```diff
@@ -0,0 +1,312 @@
+    public CqlSessionBuilder session(CassandraConfiguration configuration) {
+
+        try {
+            CqlSessionBuilder builder = CqlSession.builder().withConfigLoader(new DefaultDriverConfigLoader(() -> {
```
This is not really maintainable. We would need to find a way to handle this automatically

pollend replied:
umm, I've tried a couple different things such as creating those properties in CassandraConfiguration but they tended to be a lot more complex and cumbersome. The other option is do something like this.

```
ConfigFactory.invalidateCaches()
val ref = ConfigFactory.load().getConfig("datastax-java-driver")
val app: Config = ref.getConfig('other-prefix');
app.withFallback(ref)
```

application.conf
```
other-prefix {
    basic.config-reload-interval = 5 minutes
}
datastax-java-driver {
    basic.config-reload-interval = 5 minutes
}
```

pollend replied:
I found out I could just pass the map of properties directly.

pollend commented:
with this change you can't modify some fields such as `basic.config-reload-interval` because that takes a Duration and from what I can see I don't think this is in the yaml specification. not sure how to deal with this edge case?

willbuck commented on src/main/docs/guide/additional.adoc:
```diff
@@ -0,0 +1,25 @@
+datastax is configured using https://github.com/lightbend/config[lightbend/config] under the field `datastax-java-driver.\*`. the equivalent bean is mapped under `cassandra.*.*`. fields provided under
```
I think we could call out the full list of config options a little better here
```suggestion
The Datastax Cassandra driver is configured using https://github.com/lightbend/config[lightbend/config] under the field `datastax-java-driver.\*`. 

The equivalent bean created for micronaut is mapped under the `cassandra.*.*`. fields provided in your `application.conf`, but the bean will fallback on the `datastax-java-driver.\*` values if they are present.

These config values are provided to the   https://docs.datastax.com/en/drivers/java/4.4/com/datastax/oss/driver/api/core/CqlSession.html[CqlSession] under the hood. Datastax does not provide a class that maps to all the various possible config keys, but a full list of values that can be provided can be found in the https://github.com/datastax/java-driver/blob/4.x/core/src/main/java/com/datastax/oss/driver/api/core/config/DefaultDriverOption.java[DefaultDriverOption enum].
```

willbuck commented on src/main/docs/guide/additional.adoc:
```diff
@@ -0,0 +1,25 @@
+datastax is configured using https://github.com/lightbend/config[lightbend/config] under the field `datastax-java-driver.\*`. the equivalent bean is mapped under `cassandra.*.*`. fields provided under
+application.conf will act as fallback properties when building https://docs.datastax.com/en/drivers/java/4.4/com/datastax/oss/driver/api/core/CqlSession.html[CqlSession].
```
Going along w/ the suggested change above, this line would be replaced.
```suggestion

```

willbuck commented:
Thanks very much for all your work on this @pollend :) I think we're about ready to merge this, just had a couple more suggestions for clarifying in the docs how users can find the full list of config values to supply (since our annotation processor can't read that `DefaultDriverOption` enum for the types and add them nicely to the docs; it'd be cool if Datastax would map those config options with a class we could extend!).

I've gone ahead and branched master where it is right now to a 1.x branch, so that we can still provide bugfixes to users who elect to stay on the older driver version.

pollend commented:
could we write a bit of groovy to parse that out and add it to the documents?

pollend commented:
@willbuck any updates on this?

willbuck commented:
@pollend I'm not sure I know of a way to parse that `DefaultDriverOption.java` class in a really slick way. I'm sure some groovy *could* be written to do so, but I don't really have the resources to do it and it wouldn't be terribly re-usable so it's tough to justify. I think it makes more sense just to link to that file in the docs, for now, and perhaps file an issue on the driver to model the config with a Class we could extend (so that our docs tooling could properly generate that list).

Perhaps filing the issue would yield an existing parser the Datastax team is aware of or uses internally?

pollend commented:
I guess that could be worked out later in the docs anything besides that? @willbuck @jameskleeh
